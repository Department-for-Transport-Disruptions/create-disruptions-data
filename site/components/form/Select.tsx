import kebabCase from "lodash/kebabCase";
import { ReactElement, useState } from "react";
import FormElementWrapper, { FormGroupWrapper } from "./FormElementWrapper";
import { DisplayValuePair, ErrorInfo, FormBase } from "../../interfaces";
import { handleBlur } from "../../utils/formUtils";

interface SelectProps<T> extends FormBase<T> {
    defaultDisplay: string;
    selectValues: DisplayValuePair[];
    width?: string;
    updateOnChange?: boolean;
    useDefaultValue?: boolean;
}

const Select = <T extends object>({
    value,
    inputName,
    display,
    displaySize = "s",
    initialErrors = [],
    defaultDisplay,
    selectValues,
    stateUpdater,
    schema,
    width = "3/4",
    updateOnChange = false,
    useDefaultValue = false,
}: SelectProps<T>): ReactElement => {
    const [errors, setErrors] = useState<ErrorInfo[]>(initialErrors);
    const inputId = kebabCase(inputName);

    const getSelectOptions = (): JSX.Element[] => {
        const options: JSX.Element[] = [
            <option value="" disabled key="">
                {defaultDisplay}
            </option>,
        ];

        selectValues.forEach((input) => {
            options.push(
                <option value={input.value} key={`option-${input.value}`}>
                    {input.display}
                </option>,
            );
        });

        return options;
    };

    return (
        <FormGroupWrapper errorIds={[inputName]} errors={errors}>
            <div className="govuk-form-group" id={inputId}>
                <label className={`govuk-label govuk-label--${displaySize}`} htmlFor={`${inputId}-input`}>
                    {display}
                </label>
                <FormElementWrapper errors={errors} errorId={inputName} errorClass="govuk-select--error">
                    <select
                        className={`govuk-select w-${width}`}
                        name={inputName}
                        id={`${inputId}-input`}
                        defaultValue={useDefaultValue ? value ?? "" : undefined}
                        value={!useDefaultValue ? value ?? "" : undefined}
                        onBlur={
                            !updateOnChange
                                ? (e) => handleBlur(e.target.value, inputName, stateUpdater, setErrors, schema)
                                : undefined
                        }
                        onChange={
                            updateOnChange
                                ? (e) => handleBlur(e.target.value, inputName, stateUpdater, setErrors, schema)
                                : undefined
                        }
                    >
                        {getSelectOptions()}
                    </select>
                </FormElementWrapper>
            </div>
        </FormGroupWrapper>
    );
};

export default Select;
