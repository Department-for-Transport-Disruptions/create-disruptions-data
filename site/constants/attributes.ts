export const CD_SUMMARY = "summary";

export const CD_DESCRIPTION = "description";

export const CD_DISRUPTION_TYPE = "disruptionType";

export const CD_ASSOCIATED_LINK = "associatedLink";

export const CD_DISRUPTION_REASON = "disruptionReason";

export const CD_DISRUPTION_REAPEATS = "disruptionRepeats";

export const CD_NO_VALIDITY_END_DATE_TIME = "validityNoEndDateTime";

export const CD_NO_PUBLISH_END_DATE_TIME = "publishNoEndDateTime";

export const CD_VALIDITY_START_DATE_DAY = "validityStartDateDay";

export const CD_VALIDITY_START_DATE_MONTH = "validityStartDateMonth";

export const CD_VALIDITY_START_DATE_YEAR = "validityStartDateYear";

export const CD_VALIDITY_START_HOUR = "validityStartHour";

export const CD_VALIDITY_START_MINUTES = "validityStartMinute";

export const CD_VALIDITY_END_DATE_DAY = "validityEndDateDay";

export const CD_VALIDITY_END_DATE_MONTH = "validityEndDateMonth";

export const CD_VALIDITY_END_DATE_YEAR = "validityEndDateYear";

export const CD_VALIDITY_END_HOUR = "validityEndHour";

export const CD_VALIDITY_END_MINUTES = "validityEndMinute";

export const CD_PUBLISH_START_DATE_DAY = "publishStartDateDay";

export const CD_PUBLISH_START_DATE_MONTH = "publishStartDateMonth";

export const CD_PUBLISH_START_DATE_YEAR = "publishStartDateYear";

export const CD_PUBLISH_START_HOUR = "publishStartHour";

export const CD_PUBLISH_START_MINUTES = "publishStartMinute";

export const CD_PUBLISH_END_DATE_DAY = "publishEndDateDay";

export const CD_PUBLISH_END_DATE_MONTH = "publishEndDateMonth";

export const CD_PUBLISH_END_DATE_YEAR = "publishEndDateYear";

export const CD_PUBLISH_END_HOUR = "publishEndHour";

export const CD_PUBLISH_END_MINUTES = "publishEndMinute";
