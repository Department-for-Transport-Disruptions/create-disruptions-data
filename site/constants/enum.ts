export enum TransportMode {
    bus = "Bus",
    tram = "Tram",
    ferry = "Ferry",
    train = "Train",
}

export enum ConsequenceType {
    services = "Services",
    networkWide = "Network wide",
    operatorWide = "Operator wide",
    stops = "Stops",
}
