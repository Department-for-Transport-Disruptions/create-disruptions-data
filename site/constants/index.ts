import {
    MiscellaneousReason,
    PersonnelReason,
    EnvironmentReason,
    EquipmentReason,
} from "@create-disruptions-data/shared-ts/enums";

export const SUPPORT_EMAIL_ADDRESS = "bodshelpdesk@kpmg.co.uk";

export const SERVICE_EMAIL_ADDRESS = "bodshelpdesk@kpmg.co.uk";

export const SUPPORT_PHONE_NUMBER = "0800 123 1234";

export const DISRUPTION_REASONS: {
    value: MiscellaneousReason | PersonnelReason | EnvironmentReason | EquipmentReason;
    display: string;
}[] = [
    {
        value: MiscellaneousReason.roadWorks,
        display: "Road Works",
    },
    {
        value: MiscellaneousReason.vandalism,
        display: "Vandalism",
    },
    {
        value: MiscellaneousReason.routeDiversion,
        display: "Route Diversion",
    },
    {
        value: MiscellaneousReason.specialEvent,
        display: "Special Event",
    },
];

export const DISRUPTION_TYPES = ["planned", "unplanned"];

export const CD_DATE_FORMAT = "DD/MM/YYYY";
export const CONTACT_FEEDBACK_QUESTION = "Did you contact us for assistance at any point?";
export const SOLVE_FEEDBACK_QUESTION = "Did we solve your problem?";
export const HEAR_ABOUT_US_FEEDBACK_QUESTION = "How did you hear about our service?";
export const GENERAL_FEEDBACK_QUESTION = "Please let us know any feedback or suggestions for improvements you may have";

export const COOKIES_POLICY_COOKIE = "cdd-cookies-policy";

export const COOKIE_PREFERENCES_COOKIE = "cdd-cookie-preferences-set";

export const oneYearInSeconds = 31556952;

export const ID_TOKEN_COOKIE = "cdd-id-token";

export const COOKIES_DISRUPTION_INFO = "disruption-info";

export const COOKIES_DISRUPTION_ERRORS = "disruption-errors";

export const COOKIES_ADD_CONSEQUENCE_INFO = "cdd-consequence-info";

export const COOKIES_ADD_CONSEQUENCE_ERRORS = "cdd-consequence-errors";

export const CREATE_DISRUPTION_PAGE_PATH = "/create-disruption";

export const ERROR_PATH = "/_error";

export const ADD_CONSEQUENCE_PAGE_PATH = "/type-of-consequence";

export const STAGE = process.env.STAGE || "dev";
