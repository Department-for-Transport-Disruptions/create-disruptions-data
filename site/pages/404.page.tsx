import Link from "next/link";
import { ReactElement } from "react";
import { TwoThirdsLayout } from "../components/layout/Layout";

const title = "Error - Create Transport Disruption Data Service";
const description = "404 page for the Create Transport Disruption Data Service";

const Error = (): ReactElement => (
    <TwoThirdsLayout title={title} description={description}>
        <div>
            <h1 className="govuk-heading-l">Page not found</h1>
            <p className="govuk-body">If you typed the web address, check it is correct.</p>
            <p className="govuk-body">If you pasted the web address, check you copied the entire address.</p>
            <p className="govuk-body">
                If the web address is correct or you selected a link or button,{" "}
                <Link className="govuk-link" id="contact-link" href="/contact">
                    contact
                </Link>{" "}
                us about your disruptions data.
            </p>
        </div>

        <br />
        <Link href="/" role="button" draggable="false" className="govuk-button" data-module="govuk-button">
            Start again
        </Link>
    </TwoThirdsLayout>
);

export default Error;
