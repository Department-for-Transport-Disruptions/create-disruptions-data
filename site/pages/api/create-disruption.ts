import {
    MiscellaneousReason,
    PersonnelReason,
    EnvironmentReason,
    EquipmentReason,
} from "@create-disruptions-data/shared-ts/siriTypes";
import { NextApiRequest, NextApiResponse } from "next";
import {
    COOKIES_DISRUPTION_INFO,
    COOKIES_DISRUPTION_ERRORS,
    TEN_SECONDS_IN_MILLISECONDS,
    CREATE_DISRUPTION_PAGE_PATH,
    ERROR_PATH,
} from "../../constants/index";
import { ErrorInfo } from "../../interfaces";
import { setCookieOnResponseObject } from "../../utils/apiUtils";
import {
    checkReferrer,
    validateDisruptionType,
    validateSummary,
    validateDescription,
    validateAssociatedLink,
    validateDisruptionReasons,
    validateDateTimeSection,
    getDateTime,
} from "../../utils/apiUtils/createDisruptionValidations";
import { redirectTo } from "../../utils/index";
import { PageInputs } from "../create-disruption";

const createDisruption = (req: NextApiRequest, res: NextApiResponse): void => {
    const errors: ErrorInfo[] = [];

    checkReferrer(req.headers.referer, CREATE_DISRUPTION_PAGE_PATH, ERROR_PATH, res);

    const formFields: PageInputs = req.body as PageInputs;

    const disruptionType: "planned" | "unplanned" | undefined = formFields.typeOfDisruption;

    validateDisruptionType(disruptionType, errors, "some-error-id");

    const summary = formFields.summary;

    validateSummary(summary, errors, "some-error-id");

    const description = formFields.description;

    validateDescription(description, errors, "some-error-id");

    const associatedLink = formFields["associated-link"];

    if (associatedLink) {
        validateAssociatedLink(associatedLink, errors, "some-error-id");
    }

    const disruptionReason: MiscellaneousReason | PersonnelReason | EnvironmentReason | EquipmentReason | "" =
        formFields["disruption-reason"] || "";

    validateDisruptionReasons(disruptionReason, errors, "some-error-id");

    const disruptionStartDate: Date | null = formFields["disruption-start-date"];
    const disruptionStartTime: string = formFields["disruption-start-time"];
    const disruptionEndDate: Date | null = formFields["disruption-end-date"];
    const disruptionEndTime: string = formFields["disruption-end-time"];
    const disruptionIsNoEndDateTime: string | undefined = formFields.disruptionIsNoEndDateTime;

    validateDateTimeSection(
        disruptionStartDate,
        disruptionStartTime,
        disruptionEndDate,
        disruptionEndTime,
        disruptionIsNoEndDateTime,
        errors,
        "some-error-id",
    );
    const publishStartDate: Date | null = formFields["publish-start-date"];
    const publishStartTime: string = formFields["publish-start-time"];
    const publishEndDate: Date | null = formFields["publish-end-date"];
    const publishEndTime: string = formFields["publish-end-time"];
    const publishIsNoEndDateTime: string | undefined = formFields.publishIsNoEndDateTime;

    validateDateTimeSection(
        publishStartDate,
        publishStartTime,
        publishEndDate,
        publishEndTime,
        publishIsNoEndDateTime,
        errors,
        "some-error-id",
    );

    const disruptionData: PageInputs = {
        typeOfDisruption: disruptionType,
        summary: summary,
        description: description,
        "associated-link": associatedLink,
        "disruption-reason": disruptionReason,
        "disruption-start-date": disruptionStartDate ? getDateTime(disruptionStartDate).toDate() : disruptionStartDate,
        "disruption-end-date": disruptionEndDate ? getDateTime(disruptionEndDate).toDate() : disruptionEndDate,
        "disruption-start-time": disruptionStartTime,
        "disruption-end-time": disruptionEndTime,
        disruptionIsNoEndDateTime: disruptionIsNoEndDateTime,
        "publish-start-date": publishStartDate ? getDateTime(publishStartDate).toDate() : publishStartDate,
        "publish-end-date": publishEndDate ? getDateTime(publishEndDate).toDate() : publishEndDate,
        "publish-start-time": publishStartTime,
        "publish-end-time": publishEndTime,
        publishIsNoEndDateTime: publishIsNoEndDateTime,
        disruptionRepeats: formFields.disruptionRepeats,
    };

    setCookieOnResponseObject(
        COOKIES_DISRUPTION_INFO,
        JSON.stringify(disruptionData),
        res,
        TEN_SECONDS_IN_MILLISECONDS,
        false,
    );

    if (errors.length == 0) {
        redirectTo(res, "/");
    } else {
        setCookieOnResponseObject(
            COOKIES_DISRUPTION_ERRORS,
            JSON.stringify(errors),
            res,
            TEN_SECONDS_IN_MILLISECONDS,
            false,
        );
        redirectTo(res, CREATE_DISRUPTION_PAGE_PATH);
    }
};

export default createDisruption;
