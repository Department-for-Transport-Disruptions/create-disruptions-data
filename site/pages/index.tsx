import Link from "next/link";
import { BaseLayout } from "../components/layout/Layout";

const title = "Create Transport Disruption Data";
const description =
    "Create Transport Disruption Data is a service that allows you to generate disruptions data in SIRI-SX format";

const Home = () => {
    return (
        <BaseLayout title={title} description={description}>
            <h1 className="govuk-heading-xl">Create transport disruption data</h1>

            <p className="govuk-body">
                This service is for creating disruption data for public transport services, excluding rail, in England.
            </p>

            <p className="govuk-body">The service can be used by:</p>
            <ul className="govuk-list govuk-list--bullet">
                <li>Local transport authorities that operate their own services</li>
                <li>Bus operators running commercial services in England</li>
                <li>Local transport authorities acting on behalf of bus operators</li>
            </ul>

            <p className="govuk-body">Use this service to:</p>
            <ul className="govuk-list govuk-list--bullet">
                <li>Generate disruption data output format for a new or existing public transport disruption</li>
                <li>Download your own transport disruptions data</li>
            </ul>

            <Link
                href="/create-disruption"
                role="button"
                draggable="false"
                className="govuk-button govuk-button--start"
                data-module="govuk-button"
                id="start-now-button"
            >
                Start now
                <svg
                    className="govuk-button__start-icon"
                    xmlns="http://www.w3.org/2000/svg"
                    width="17.5"
                    height="19"
                    viewBox="0 0 33 40"
                    role="presentation"
                    focusable="false"
                >
                    <path fill="currentColor" d="M0 0h13l20 20-20 20H0l20-20z" />
                </svg>
            </Link>
        </BaseLayout>
    );
};

export const getServerSideProps = (): { props: object } => ({
    props: {},
});

export default Home;
