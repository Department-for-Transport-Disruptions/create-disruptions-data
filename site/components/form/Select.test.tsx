import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import renderer from "react-test-renderer";
import { describe, it, expect, vi } from "vitest";
import Select from "./Select";
import { DISRUPTION_REASONS } from "../../constants";
import { PageInputs } from "../../pages/create-disruption";

describe("Select", () => {
    it("should render correctly with no errors", () => {
        const tree = renderer
            .create(
                <Select<PageInputs>
                    inputId="disruption-reason"
                    inputName="disruptionReason"
                    display="Reason for disruption"
                    defaultDisplay="Select a reason"
                    errorMessage="Select a reason from the dropdown"
                    selectValues={DISRUPTION_REASONS}
                    stateUpdater={vi.fn()}
                    value={""}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should render correctly with errors", () => {
        const tree = renderer
            .create(
                <Select<PageInputs>
                    inputId="disruption-reason"
                    inputName="disruptionReason"
                    display="Reason for disruption"
                    defaultDisplay="Select a reason"
                    errorMessage="Select a reason from the dropdown"
                    selectValues={DISRUPTION_REASONS}
                    stateUpdater={vi.fn()}
                    value={""}
                    initialErrors={[{ errorMessage: "There was an error", id: "disruption-reason" }]}
                />,
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("should validate minLength and display error", async () => {
        const { unmount } = render(
            <Select<PageInputs>
                inputId="disruption-reason"
                inputName="disruptionReason"
                display="Reason for disruption"
                defaultDisplay="Select a reason"
                errorMessage="Select a reason from the dropdown"
                selectValues={DISRUPTION_REASONS}
                stateUpdater={vi.fn()}
                value={""}
            />,
        );

        await userEvent.click(screen.getByText("Select a reason"));
        await userEvent.tab();

        expect(screen.getByText("Select a reason")).toBeTruthy();

        unmount();
    });
});
