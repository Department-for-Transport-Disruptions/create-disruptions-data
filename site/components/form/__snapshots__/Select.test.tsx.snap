// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Select > should render correctly with errors 1`] = `
<div
  className="govuk-form-group govuk-form-group--error"
>
  <div
    className="govuk-form-group"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="disruption-reason"
    >
      Reason for disruption
    </label>
    <div
      className=""
    >
      <span
        className="govuk-error-message"
        id="disruption-reason-error"
      >
        <span
          className="govuk-visually-hidden"
        >
          Error: 
        </span>
        There was an error
      </span>
      <select
        aria-describedby="disruption-reason-error"
        className="govuk-select w-3/4 govuk-select--error"
        defaultValue=""
        id="disruption-reason"
        name="disruptionReason"
        onBlur={[Function]}
      >
        <option
          disabled={true}
          value=""
        >
          Select a reason
        </option>
        <option
          value="roadWorks"
        >
          Road Works
        </option>
        <option
          value="vandalism"
        >
          Vandalism
        </option>
        <option
          value="routeDiversion"
        >
          Route Diversion
        </option>
        <option
          value="specialEvent"
        >
          Special Event
        </option>
      </select>
    </div>
  </div>
</div>
`;

exports[`Select > should render correctly with no errors 1`] = `
<div
  className="govuk-form-group"
>
  <div
    className="govuk-form-group"
  >
    <label
      className="govuk-label govuk-label--s"
      htmlFor="disruption-reason"
    >
      Reason for disruption
    </label>
    <div
      className=""
    >
      <select
        className="govuk-select w-3/4"
        defaultValue=""
        id="disruption-reason"
        name="disruptionReason"
        onBlur={[Function]}
      >
        <option
          disabled={true}
          value=""
        >
          Select a reason
        </option>
        <option
          value="roadWorks"
        >
          Road Works
        </option>
        <option
          value="vandalism"
        >
          Vandalism
        </option>
        <option
          value="routeDiversion"
        >
          Route Diversion
        </option>
        <option
          value="specialEvent"
        >
          Special Event
        </option>
      </select>
    </div>
  </div>
</div>
`;
